File: 3445.maximum-difference-between-even-and-odd-frequency-ii.py
#
# @lc app=leetcode id=3445 lang=python3
#
# [3445] Maximum Difference Between Even and Odd Frequency II
#

# @lc code=start
class Solution:
    def maxDifference(self, s: str, k: int) -> int:
        from collections import Counter
        n = len(s)
        max_diff = float('-inf')
        digits = ['0', '1', '2', '3', '4']
        for window_size in range(k, n+1):
            freq = Counter(s[:window_size])
            # Evaluate initial window
            for a in digits:
                if freq[a] % 2 == 1:
                    for b in digits:
                        if a != b and freq[b] > 0 and freq[b] % 2 == 0:
                            diff = freq[a] - freq[b]
                            max_diff = max(max_diff, diff)
            # Slide window
            for i in range(window_size, n):
                out_char = s[i-window_size]
                in_char = s[i]
                freq[out_char] -= 1
                freq[in_char] += 1
                for a in digits:
                    if freq[a] % 2 == 1:
                        for b in digits:
                            if a != b and freq[b] > 0 and freq[b] % 2 == 0:
                                diff = freq[a] - freq[b]
                                max_diff = max(max_diff, diff)
        return max_diff if max_diff != float('-inf') else -1
# @lc code=end

File: 3445.maximum-difference-between-even-and-odd-frequency-ii.java
#
# @lc app=leetcode id=3445 lang=java
#
# [3445] Maximum Difference Between Even and Odd Frequency II
#

# @lc code=start
class Solution {
    public int maxDifference(String s, int k) {
        int n = s.length();
        int maxDiff = Integer.MIN_VALUE;
        // Only 5 digits: '0'-'4'
        for (int left = 0; left <= n-k; ++left) {
            int[] freq = new int[5];
            for (int right = left; right < n; ++right) {
                freq[s.charAt(right)-'0']++;
                if (right-left+1 >= k) {
                    // For every (a, b)
                    for (int a = 0; a < 5; ++a) {
                        if (freq[a]%2 == 1) { // odd freq
                            for (int b = 0; b < 5; ++b) {
                                if (a != b && freq[b] > 0 && freq[b]%2 == 0) { // even, nonzero
                                    maxDiff = Math.max(maxDiff, freq[a] - freq[b]);
                                }
                            }
                        }
                    }
                }
            }
        }
        return maxDiff == Integer.MIN_VALUE ? -1 : maxDiff;
    }
}
# @lc code=end